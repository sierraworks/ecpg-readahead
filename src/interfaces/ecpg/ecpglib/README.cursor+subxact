Theory of operation for client side cursor accounting and readahead
===================================================================

Contents of tuples returned by a cursor always reflect the data present at
the time of declaring the cursor. Because of this, executing DML commands
involving the table(s) appearing in the cursor's SELECT query won't modify
the result set returned by the cursor. Because of this, cursor readahead is
possible and the cursor readahead window always contains the same data as
the underlying cursor.

Tracking the maximum known backend position value allows the number of MOVE
commands sent to the server to be reduced and re-fetching the same readahead
window avoided in case the application suddenly changes cursor positioning
from positive to negative and vice-versa.

The idea behind cursor readahead is to move one tuple less than asked by
the application and then ecpglib performs a FETCH FORWARD/BACKWARD N to
populate the local tuple cache.

By using a readahead window larger than 1 tuple, the cursor positions known
by the backend and the application are different most of the time. As a
consequence, UPDATE/DELETE ... WHERE CURRENT OF requires a MOVE behind the
application's back. This may reduce or eliminate the caching effects in the
cursor and it will not work for non-scrollable cursors. Either the cursor
must be scrollable or READAHEAD 1 must be explicitly specified for such
cursors.

The slight problem with this approach is that ABSOLUTE and RELATIVE movements
needs two operations. If the next command issued by the application spills
out of the cache and it happens frequently, it may actually decrease
performance. To prevent this, the ECPG runtime switches to executing these
kinds of commands as is after 3 cache misses. FETCH FORWARD/BACKWARD allows
switching back to caching.

The ECPG runtime library has "struct subxact_descriptor" and
"struct cursor_descriptor" for tracking subtransactions and cursors.

Explanation of structure members of "struct cursor_descriptor" are below.

subxact_level: The cursor was declared in this transaction level.
--------------
Valid values are:
0:  Outside of transactions. This is a WITH HOLD cursor, committed
    in a previous transaction.
1:  The cursor was declared in the currently running toplevel
    transaction.
>2: The cursor was declared in the currently running transaction
    in the specified SAVEPOINT.

As soon as a (sub-)transaction is committed or rolled back, cursors are
either implicitly propagated to the next subtransaction level upwards,
i.e. subxact_level is decreased by 1.

with_hold: WITH HOLD cursor (true or false)
----------

scrollable: The scrollability of the cursor.
-----------
Accepted values are: ECPGcs_no_scroll or ECPGcs_scroll

There is also a third enum value, ECPGcs_unspecified which is used
internally by the ECPG preprocessor when the application didn't
explicitly specify neither SCROLL nor NO SCROLL. It will be eventually
overridden by NO SCROLL in the DECLARE statement and ECPGcs_no_scroll
in the "scrollable" argument.

res: The current readahead window.
----
It is used to return data across several FETCH commands, speeding up
the application by reducing network turnaround.

readahead: The size of the readahead window.
----------
The ECPG runtime library attempts to fetch this amount of records at once.

cache_start_pos: The start position of the readahead window.
----------------
The value is 1-based. Position 0 means that the cursor is before the first
tuple in the result set.

I can also be negative (but also 1-based) if the application started with
FETCH/MOVE ABSOLUTE <some negative number>. In this case the numbering
is used to count from the end of the result set.

backward: The ECPG runtime library fetched this readahead window backward.
---------

atstart: The current position is at the start of the result set.
--------

atend: The current position is at the end of the result set.
------

pos: The current cursor position, as known by the application.
----
The comments above for "cache_start_pos" are also applicable here.
"atstart" and "atend" apply to this position.

backend_pos: The current cursor position, as known by the server.
------------

mkbpos: The maximum known backend position.
-------
If the application tries to execute a MOVE and it's over the previously
known maximum value, different things happen depending on the type of the
motion.

MOVE [FORWARD|BACKWARD] ALL
The number of records from the previous position to the end of the result
set is reported.

MOVE ABSOLUTE N, MOVE [FORWARD|BACKWARD] RELATIVE N
If there is a tuple on this position, the server indicates the existence of
a tuple on that position.

This value is trivially updated for FETCH commands.

mkbp_is_last: Maximum Known Backend Position is the last valid position.
-------------
The "mkbpos" is the last cursor position where a tuple exists 

Structure members of "struct subxact_descriptor" don't need too much
explanation. New savepoints get prepended to the list of subtransations
which are per-connection, so the head of the list is found in
struct connection. However, the necessity of this structure is explained
below.

Tracking (sub-)transactions
===========================

One side effect/requirement of tracking cursors in the ECPG runtime library
is that (sub-)transactions are also needed to be tracked. These two are
in some kind of symbiosis: if a command fails, the (sub-)transaction returns
"current transaction is aborted" error for every subsequent commands.
Such a failed command can be a cursor command using an invalid cursor name.

By tracking cursors in the ECPG runtime library, detecting an invalid cursor
name can be done early which reduces the network turnaround further.

But doing this, the server and the application now disagree on the failed
transaction. Because of this, a "client side error" flag was introduced into
"struct connection".

Subsequent commands (in execute.c::ecpg_prologue() and misc.c::ECPGtrans())
check this flag and return the "current transaction is aborted" error to the
application without executing them in the server. As mentioned above, the
server and the application at this point disagree on the failed transaction
and he issued command may be executed without an error, causing unwanted
effects.

Checking the "client side error" flag further reduces network turnaround for
transactions failed on the client side and buggy applications that don't check
for failed commands and roll back the (sub-)transaction.

Executing COMMIT or PREPARE TRANSACTION for a failed transaction silently
overrides these commands as ROLLBACK in the server. By checking the "client
side error" flag, it can be done explicitly.

In a failed transaction, only the ROLLBACK and ROLLBACK TO SAVEPOINT commands
are allowed to be executed without errors in the server. These also pass
through in client side tracking code, allowing the "client side error" flag
to be cleared.

All other transaction commands are treated as regular commands in a failed
transactions in that the "current transaction is aborted" error is returned.

Cursors are mostly transactional, i.e. rolling back a (sub-)transaction makes
cursors declared in that subtransaction forgotten in the server.

Committing a transaction or releasing a subtransaction make cursors declared
in this (sub-)transaction or some levels below that propagates the cursor
upwards with only WITH HOLD cursors surviving a committed transaction.

The ECPG runtime library must also follow the server behaviour in tracking
cursors across (sub-)transactions. This is done in misc.c in the ECPGtrans()
function and in cursor.c in the ecpg_commit_cursors() function.
