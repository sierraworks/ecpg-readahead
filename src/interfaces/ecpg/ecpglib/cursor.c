/* src/interfaces/ecpg/ecpglib/cursor.c */

/*
 * ECPG client-side cursor support routines
 */

#define POSTGRES_ECPG_INTERNAL
#include "postgres_fe.h"

#include <stdlib.h>
#include <limits.h>

#include "ecpgtype.h"
#include "ecpglib.h"
#include "ecpgerrno.h"
#include "extern.h"

static void
raise_cursor_error(struct connection *con, int lineno, const char *name)
{
	con->client_side_error = true;
	ecpg_raise(lineno, ECPG_INVALID_CURSOR,
						ECPG_SQLSTATE_INVALID_CURSOR_NAME,
						(name ? name : ecpg_gettext("<empty>")));
}

/*
 * Find the cursor with the given name.
 * Optionally, return the previous one in the list.
 */
static bool
find_cursor(struct connection *con, int lineno, const char *name, bool ignore,
				struct cursor_descriptor **out,
				struct cursor_descriptor **out_prev)
{
	struct cursor_descriptor *desc = con->cursor_desc,
					*prev = NULL;
	int (* strcmp_fn)(const char *, const char *);
	bool	found = false;

	if (!name)
	{
		raise_cursor_error(con, lineno, name);
		if (out)
			*out = NULL;
		if (out_prev)
			*out_prev = NULL;
		return false;
	}

	/*
	 * This will ensure that:
	 * - quoted cursor names will be the first ones in the list ordered case-sensitively
	 * - unquoted cursor names at the end, ordered case-insensitively.
	 */
	strcmp_fn = (name[0] == '"' ? strcmp : pg_strcasecmp);

	while (desc)
	{
		int ret = strcmp_fn(desc->name, name);

		if (ret == 0)
		{
			found = true;
			break;
		}
		/*
		 * The cursor list is ordered quasi alphabetically,
		 * don't search the whole list.
		 */
		if (ret > 0)
			break;
		prev = desc;
		desc = desc->next;
	}

	if (!found)
	{
		if (!ignore)
			raise_cursor_error(con, lineno, name);
		desc = NULL;
	}

	if (out)
		*out = desc;
	if (out_prev)
		*out_prev = prev;
	return found;
}

/*
 * Add a new cursor descriptor, maintain "mostly" alphabetic order.
 * - case sensitive alphabetic order for quoted cursor names
 * - case insensitive alphabetic order for non-quoted cursor names
 */
static struct cursor_descriptor *
add_cursor(int lineno, struct connection *con,
					const char *name,
					int subxact_level,
					bool with_hold)
{
	struct cursor_descriptor *desc, *prev;

	if (!name || name[0] == '\0')
		return NULL;

	if (find_cursor(con, lineno, name, true, &desc, &prev))
		return desc;

	desc = (struct cursor_descriptor *)
			ecpg_alloc(sizeof(struct cursor_descriptor),
								lineno);
	if (!desc)
		return NULL;

	/* Copy the name */
	desc->name = ecpg_strdup(name, lineno);
	if (!desc->name)
	{
		ecpg_free(desc);
		return NULL;
	}

	desc->subxact_level = subxact_level;
	desc->with_hold = with_hold;

	if (prev)
	{
		desc->next = prev->next;
		prev->next = desc;
	}
	else
	{
		desc->next = con->cursor_desc;
		con->cursor_desc = desc;
	}

	return desc;
}

/*
 * Delete the cursor from the list of descriptors.
 */
static void
del_cursor(struct connection *con, int lineno, const char *name)
{
	struct cursor_descriptor *desc, *prev;

	if (find_cursor(con, lineno, name, true, &desc, &prev))
	{
		if (prev)
			prev->next = desc->next;
		else
			con->cursor_desc = desc->next;

		ecpg_free((void *)desc->name);
		ecpg_free(desc);
	}
}

/*
 * Commit or roll back cursors:
 * - all transactional (non-WITH HOLD) cursors are forgotten at
 *   the end of the transaction, no matter if COMMIT or ROLLBACK
 * - at RELEASE SAVEPOINT, the cursors are propagated one
 *   subtransaction level up.
 * - at ROLLBACK TO SAVEPOINT, all cursors are forgotten that were
 *   created in this subtransaction or below that level
 * - WITH HOLD cursors are kept after the transaction in case of
 *   toplevel COMMIT. The caller checks for the transaction in error
 *   and sets the rollback flag accordingly.
 */
void
ecpg_commit_cursors(int lineno, struct connection * conn,
						bool rollback, int level)
{
	struct cursor_descriptor *desc, *tmp,
					*prev = NULL;

	desc = conn->cursor_desc;
	while (desc)
	{
		/* Keep already COMMITted WITH HOLD cursors */
		if (desc->with_hold && desc->subxact_level == 0)
			;
		else if (rollback)
		{
			/*
			 * On ROLLBACK [ TO SAVEPOINT ... ] forget about
			 * every cursor DECLAREd under this level.
			 */
			if (desc->subxact_level >= level)
			{
				tmp = desc->next;
				ecpg_free(desc->name);
				ecpg_free(desc);
				desc = tmp;
				if (prev)
					prev->next = tmp;
				else
					conn->cursor_desc = tmp;
				continue;
			}
		}
		else
		{
			/*
			 * If this is a WITH HOLD cursor or we're in a
			 * subtransaction, COMMIT the cursor one level up.
			 */
			if (desc->with_hold || level > 1)
			{
				if (level > 0)
					desc->subxact_level = level - 1;
			}
			/*
			 * Otherwise (not WITH HOLD cursor AND it's the
			 * toplevel transaction) free the cursor.
			 */
			else
			{
				tmp = desc->next;
				ecpg_free(desc->name);
				ecpg_free(desc);
				desc = tmp;
				if (prev)
					prev->next = tmp;
				else
					conn->cursor_desc = tmp;
				continue;
			}
		}
		prev = desc;
		desc = desc->next;
	}
}

/*
 * Declare the cursor in the backend and create a
 * minimalistic descriptor to make tracking possible.
 * This function is called when the ECPG code OPENs the cursor.
 */
bool
ECPGopen(const int lineno, const int compat, const int force_indicator,
			const char *connection_name, const bool questionmarks,
			const bool with_hold, const char *curname, const int st, const char *query, ...)
{
	struct connection *con = ecpg_get_connection(connection_name);
	int		subxact_level;
	bool		ret;
	va_list		args;

	/*
	 * Execute the DECLARE query in the backend.
	 */
	va_start(args, query);
	ret = ecpg_do(lineno, compat, force_indicator, connection_name, questionmarks,
										st,
										query,
										args);
	va_end(args);

	if (!ret)
		return false;

	/* Finally add the cursor to the connection. */
	if (con->subxact_desc)
		subxact_level = con->subxact_desc->level;
	else
		subxact_level =
			(PQtransactionStatus(con->connection) != PQTRANS_IDLE ?
										1 : 0);
	add_cursor(lineno, con, curname, subxact_level, with_hold);

	return true;
}

/*
 * ECPGfetch
 * Execute a FETCH or MOVE statement for the application.
 *
 * This function maintains the internal cursor descriptor and
 * reduces the the network roundtrip by returning early for
 * an unknown cursor name.
 */
bool
ECPGfetch(const int lineno, const int compat, const int force_indicator,
				const char *connection_name, const bool questionmarks,
				const char *curname, const int st, const char *query, ...)
{
	struct connection  *con = ecpg_get_connection(connection_name);
	struct cursor_descriptor *cur;
	bool		ret;
	va_list		args;

	if (!find_cursor(con, lineno, curname, false, &cur, NULL))
		return false;

	va_start(args, query);
	ret = ecpg_do(lineno, compat, force_indicator, connection_name, questionmarks,
									st,
									query,
									args);
	va_end(args);

	return ret;
}

/*
 * ECPGcursor_dml
 * Called for DELETE / UPDATE ... WHERE CURRENT OF cursorname
 */
bool
ECPGcursor_dml(const int lineno, const int compat, const int force_indicator,
				const char *connection_name, const bool questionmarks,
				const char *curname, const int st, const char *query, ...)
{
	struct connection  *con = ecpg_get_connection(connection_name);
	struct cursor_descriptor *cur;
	bool		ret;
	va_list		args;

	if (!find_cursor(con, lineno, curname, false, &cur, NULL))
		return false;

	va_start(args, query);
	ret = ecpg_do(lineno, compat, force_indicator, connection_name, questionmarks,
									st, query, args);
	va_end(args);
	return ret;
}

/*
 * ECPGclose
 * Close and free a cursor.
 */
bool
ECPGclose(const int lineno, const int compat, const int force_indicator,
				const char *connection_name, const bool questionmarks,
				const char *curname, const int st, const char *query, ...)
{
	struct connection  *con = ecpg_get_connection(connection_name);
	struct cursor_descriptor *cur;
	bool		ret;
	va_list		args;

	if (!find_cursor(con, lineno, curname, false, &cur, NULL))
		return false;

	va_start(args, query);
	ret = ecpg_do(lineno, compat, force_indicator, connection_name, questionmarks,
									st, query, args);
	va_end(args);

	del_cursor(con, lineno, curname);

	return ret;
}
