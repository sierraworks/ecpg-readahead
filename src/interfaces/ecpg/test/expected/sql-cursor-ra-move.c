/* Processed by ecpg (regression mode) */
/* These include files are added by the preprocessor */
#include <ecpglib.h>
#include <ecpgerrno.h>
#include <sqlca.h>
/* End of automatic include section */
#define ECPGdebug(X,Y) ECPGdebug((X)+100,(Y))

#line 1 "cursor-ra-move.pgc"
#include <stdlib.h>
#include <string.h>


#line 1 "regression.h"






#line 4 "cursor-ra-move.pgc"


/* exec sql whenever sqlerror  sqlprint ; */
#line 6 "cursor-ra-move.pgc"


int main(void)
{
	int	quit_loop;

	ECPGdebug(1, stderr);

	{ ECPGconnect(__LINE__, 0, "regress1" , NULL, NULL , NULL, 0); 
#line 14 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 14 "cursor-ra-move.pgc"


	{ ECPGdo(__LINE__, 0, 1, NULL, 0, ECPGst_normal, "create table t1 ( id serial primary key , t text )", ECPGt_EOIT, ECPGt_EORT);
#line 16 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 16 "cursor-ra-move.pgc"

	{ ECPGdo(__LINE__, 0, 1, NULL, 0, ECPGst_normal, "insert into t1 ( t ) values ( 'a' ) , ( 'b' ) , ( 'c' ) , ( 'd' ) , ( 'e' ) , ( 'f' ) , ( 'g' ) , ( 'h' ) , ( 'i' ) , ( 'j' ) , ( 'k' ) , ( 'l' ) , ( 'm' ) , ( 'n' ) , ( 'o' ) , ( 'p' ) , ( 'q' ) , ( 'r' ) , ( 's' ) , ( 't' ) , ( 'u' ) , ( 'v' ) , ( 'w' ) , ( 'x' ) , ( 'y' ) , ( 'z' )", ECPGt_EOIT, ECPGt_EORT);
#line 21 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 21 "cursor-ra-move.pgc"

	{ ECPGtrans(__LINE__, NULL, "commit", 0, 0, 0, NULL);
#line 22 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 22 "cursor-ra-move.pgc"


	/* declare scroll_cur scroll cursor for select id , t from t1 order by id */
#line 24 "cursor-ra-move.pgc"

	/* declare noscroll_cur no scroll cursor for select id , t from t1 order by id */
#line 25 "cursor-ra-move.pgc"

	/* Implicitly non-scrollable cursors */
	/* declare unspec_cur1 no scroll cursor for select id , t from t1 order by id */
#line 27 "cursor-ra-move.pgc"

	/* declare unspec_cur2 no scroll cursor for select t1 . id , t1 . t , t2 . id , t2 . t from t1 join t1 as t2 on ( t1 . id = 27 - t2 . id ) order by t1 . id */
#line 28 "cursor-ra-move.pgc"


	/* Test MOVE ABSOLUTE -1 for all 4 cursors */

	printf("test scroll_cur for move absolute -1\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_scroll, 8, 0, 0, "scroll_cur", ECPGst_normal, "declare scroll_cur scroll cursor for select id , t from t1 order by id", ECPGt_EOIT, ECPGt_EORT);
#line 34 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 34 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("open failed with SQLSTATE %5s\n", sqlca.sqlstate);

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_absolute, "- 1", 1, "scroll_cur", ECPGst_normal, "move absolute - 1 in scroll_cur", ECPGt_EOIT, ECPGt_EORT);
#line 38 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 38 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("move absolute -1 in scroll_cur failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("move absolute -1 in scroll_cur succeeded\n");

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 44 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 44 "cursor-ra-move.pgc"


	printf("test noscroll_cur for move absolute -1\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 8, 0, 0, "noscroll_cur", ECPGst_normal, "declare noscroll_cur no scroll cursor for select id , t from t1 order by id", ECPGt_EOIT, ECPGt_EORT);
#line 48 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 48 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("open failed with SQLSTATE %5s\n", sqlca.sqlstate);

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_absolute, "- 1", 1, "noscroll_cur", ECPGst_normal, "move absolute - 1 in noscroll_cur", ECPGt_EOIT, ECPGt_EORT);
#line 52 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 52 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("move absolute -1 in noscroll_cur failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("move absolute -1 in noscroll_cur succeeded\n");

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 58 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 58 "cursor-ra-move.pgc"


	printf("test unspec_cur1 for move absolute -1\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 8, 0, 0, "unspec_cur1", ECPGst_normal, "declare unspec_cur1 no scroll cursor for select id , t from t1 order by id", ECPGt_EOIT, ECPGt_EORT);
#line 62 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 62 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("open failed with SQLSTATE %5s\n", sqlca.sqlstate);

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_absolute, "- 1", 1, "unspec_cur1", ECPGst_normal, "move absolute - 1 in unspec_cur1", ECPGt_EOIT, ECPGt_EORT);
#line 66 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 66 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("move absolute -1 in unspec_cur1 failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("move absolute -1 in unspec_cur1 succeeded\n");

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 72 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 72 "cursor-ra-move.pgc"


	printf("test unspec_cur2 for move absolute -1\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_no_scroll, 8, 0, 0, "unspec_cur2", ECPGst_normal, "declare unspec_cur2 no scroll cursor for select t1 . id , t1 . t , t2 . id , t2 . t from t1 join t1 as t2 on ( t1 . id = 27 - t2 . id ) order by t1 . id", ECPGt_EOIT, ECPGt_EORT);
#line 76 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 76 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("open failed with SQLSTATE %5s\n", sqlca.sqlstate);

	{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_absolute, "- 1", 1, "unspec_cur2", ECPGst_normal, "move absolute - 1 in unspec_cur2", ECPGt_EOIT, ECPGt_EORT);
#line 80 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 80 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("move absolute -1 in unspec_cur2 failed with SQLSTATE %5s\n", sqlca.sqlstate);
	else
		printf("move absolute -1 in unspec_cur2 succeeded\n");

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 86 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 86 "cursor-ra-move.pgc"


	/* Test MOVE RELATIVE 8 */

	printf("test scroll_cur for move relative 8\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_scroll, 8, 0, 0, "scroll_cur", ECPGst_normal, "declare scroll_cur scroll cursor for select id , t from t1 order by id", ECPGt_EOIT, ECPGt_EORT);
#line 92 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 92 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("open failed with SQLSTATE %5s\n", sqlca.sqlstate);

	quit_loop = 0;
	while (!quit_loop)
	{
		{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_relative, "8", 1, "scroll_cur", ECPGst_normal, "move relative 8 in scroll_cur", ECPGt_EOIT, ECPGt_EORT);
#line 99 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 99 "cursor-ra-move.pgc"

		if (sqlca.sqlcode < 0)
			printf("move relative 8 in scroll_cur failed with SQLSTATE %5s\n", sqlca.sqlstate);
		else
			printf("move relative 8 in scroll_cur succeeded, sqlerrd[2] %ld\n", sqlca.sqlerrd[2]);
		quit_loop = (sqlca.sqlerrd[2] == 0);
	}

	printf("test scroll_cur for move relative -8\n");

	quit_loop = 0;
	while (!quit_loop)
	{
		{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_relative, "- 8", 1, "scroll_cur", ECPGst_normal, "move relative - 8 in scroll_cur", ECPGt_EOIT, ECPGt_EORT);
#line 112 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 112 "cursor-ra-move.pgc"

		if (sqlca.sqlcode < 0)
			printf("move relative -8 in scroll_cur failed with SQLSTATE %5s\n", sqlca.sqlstate);
		else
			printf("move relative -8 in scroll_cur succeeded, sqlerrd[2] %ld\n", sqlca.sqlerrd[2]);
		quit_loop = (sqlca.sqlerrd[2] == 0);
	}

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 120 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 120 "cursor-ra-move.pgc"


	/* Test MOVE FORWARD 8 */

	printf("test scroll_cur for move forward 8\n");

	{ ECPGopen(__LINE__, 0, 1, NULL, 0, 0, ECPGcs_scroll, 8, 0, 0, "scroll_cur", ECPGst_normal, "declare scroll_cur scroll cursor for select id , t from t1 order by id", ECPGt_EOIT, ECPGt_EORT);
#line 126 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 126 "cursor-ra-move.pgc"

	if (sqlca.sqlcode < 0)
		printf("open failed with SQLSTATE %5s\n", sqlca.sqlstate);

	quit_loop = 0;
	while (!quit_loop)
	{
		{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_forward, "8", 1, "scroll_cur", ECPGst_normal, "move forward 8 in scroll_cur", ECPGt_EOIT, ECPGt_EORT);
#line 133 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 133 "cursor-ra-move.pgc"

		if (sqlca.sqlcode < 0)
			printf("move forward 8 in scroll_cur failed with SQLSTATE %5s\n", sqlca.sqlstate);
		else
			printf("move forward 8 in scroll_cur succeeded, sqlerrd[2] %ld\n", sqlca.sqlerrd[2]);
		quit_loop = (sqlca.sqlerrd[2] == 0);
	}

	printf("test scroll_cur for move forward -8\n");

	quit_loop = 0;
	while (!quit_loop)
	{
		{ ECPGfetch(__LINE__, 0, 1, NULL, 0, ECPGc_forward, "- 8", 1, "scroll_cur", ECPGst_normal, "move forward - 8 in scroll_cur", ECPGt_EOIT, ECPGt_EORT);
#line 146 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 146 "cursor-ra-move.pgc"

		if (sqlca.sqlcode < 0)
			printf("move forward -8 in scroll_cur failed with SQLSTATE %5s\n", sqlca.sqlstate);
		else
			printf("move forward -8 in scroll_cur succeeded, sqlerrd[2] %ld\n", sqlca.sqlerrd[2]);
		quit_loop = (sqlca.sqlerrd[2] == 0);
	}

	{ ECPGtrans(__LINE__, NULL, "rollback", 0, 0, 1, NULL);
#line 154 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 154 "cursor-ra-move.pgc"


	/* Drop the test table */

	{ ECPGdo(__LINE__, 0, 1, NULL, 0, ECPGst_normal, "drop table t1", ECPGt_EOIT, ECPGt_EORT);
#line 158 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 158 "cursor-ra-move.pgc"

	{ ECPGtrans(__LINE__, NULL, "commit", 0, 0, 0, NULL);
#line 159 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 159 "cursor-ra-move.pgc"


	{ ECPGdisconnect(__LINE__, "ALL");
#line 161 "cursor-ra-move.pgc"

if (sqlca.sqlcode < 0) sqlprint();}
#line 161 "cursor-ra-move.pgc"


	return 0;
}
