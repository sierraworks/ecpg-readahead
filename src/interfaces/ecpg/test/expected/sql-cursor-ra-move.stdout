test scroll_cur for move absolute -1
move absolute -1 in scroll_cur succeeded
test noscroll_cur for move absolute -1
move absolute -1 in noscroll_cur failed with SQLSTATE 55000
test unspec_cur1 for move absolute -1
move absolute -1 in unspec_cur1 failed with SQLSTATE 55000
test unspec_cur2 for move absolute -1
move absolute -1 in unspec_cur2 failed with SQLSTATE 55000
test scroll_cur for move relative 8
move relative 8 in scroll_cur succeeded, sqlerrd[2] 1
move relative 8 in scroll_cur succeeded, sqlerrd[2] 1
move relative 8 in scroll_cur succeeded, sqlerrd[2] 1
move relative 8 in scroll_cur succeeded, sqlerrd[2] 0
test scroll_cur for move relative -8
move relative -8 in scroll_cur succeeded, sqlerrd[2] 1
move relative -8 in scroll_cur succeeded, sqlerrd[2] 1
move relative -8 in scroll_cur succeeded, sqlerrd[2] 1
move relative -8 in scroll_cur succeeded, sqlerrd[2] 0
test scroll_cur for move forward 8
move forward 8 in scroll_cur succeeded, sqlerrd[2] 8
move forward 8 in scroll_cur succeeded, sqlerrd[2] 8
move forward 8 in scroll_cur succeeded, sqlerrd[2] 8
move forward 8 in scroll_cur succeeded, sqlerrd[2] 2
move forward 8 in scroll_cur succeeded, sqlerrd[2] 0
test scroll_cur for move forward -8
move forward -8 in scroll_cur succeeded, sqlerrd[2] 8
move forward -8 in scroll_cur succeeded, sqlerrd[2] 8
move forward -8 in scroll_cur succeeded, sqlerrd[2] 8
move forward -8 in scroll_cur succeeded, sqlerrd[2] 2
move forward -8 in scroll_cur succeeded, sqlerrd[2] 0
